File: 025/cpp/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool full(ListNode *head, int k) {
        int count = 0;
        for (ListNode *temp = head; temp != nullptr && count < k; temp = temp->next, count++);
        return count == k;
    }

    ListNode *reverse(ListNode *head, int k) {
        ListNode *dummy = nullptr;
        for (ListNode *temp = head; k > 0; temp = temp->next, k--) {
            std::cout << temp->val << " ";
            ListNode *t = new ListNode(temp->val, dummy);
            dummy = t;
        }
        return dummy;
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *root = new ListNode();
        ListNode *temp = root;
        while (head != nullptr) {
            if (!full(head, k)) {
                temp->next = head;
                break;
            }
            temp->next = reverse(head, k);
            for (int i = 0; i < k; i++, temp = temp->next, head = head->next);
        }
        return root->next;
    }
};

File: 025/rust/main.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    fn full(mut head: Option<&ListNode>, k: i32) -> bool {
        let mut count = 0;
        while head.is_some() && count < k {
            head = head.unwrap().next.as_deref();
            count += 1;
        }
        return count == k;
    }

    fn reverse(mut head: Option<&ListNode>, k: i32) -> Option<Box<ListNode>> {
        let mut dummy = None;
        for _ in 0..k {
            let t = head.unwrap();
            let mut temp = ListNode::new(t.val);
            temp.next = dummy;
            dummy = Some(Box::new(temp));
            head = t.next.as_deref();
        }
        return dummy;
    }

    pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        let mut root = ListNode::new(0);
        let mut temp = &mut root;
        let mut t = head.as_deref();
        while t.is_some() {
            if !Solution::full(t, k) {
                temp.next = Some(Box::new((*t.unwrap()).clone()));
                break;
            } 
            temp.next = Solution::reverse(t, k);
            for _ in 0..k {
                temp = temp.next.as_deref_mut().unwrap();
                t = t.unwrap().next.as_deref();
            }
        }
        return root.next;
    }
}
